guides/17_Migration_to_3_0.md

```diff
@@ -0,0 +1,120 @@
+removeObserver(observer: EventObserver): void;
+forEachObserver(observerFunc: (observer: EventObserver) => void): void;
+```
+
+Additionally because of this the constructor of `DefaultEventController` has been updated
+```typescript
+// Before in 2.0:
+constructor(
+    audioVideoController: AudioVideoController
+    eventReporter?: EventReporter
+  )
+
+// After in 3.0:
+constructor(
+    configuration: MeetingSessionConfiguration,
+    logger: Logger,
+    eventReporter?: EventReporter
+  )
+```
+
+### Updating `EventReporter` use cases
+
+The `DefaultMeetingSession` constructor no longer takes in a `EventReporter`.
+
+Before in 2.0:
+```typescript
+const configuration = new MeetingSessionConfiguration(…);
+const eventReporter = new EventReporter(...)
+…
+const deviceController = new DefaultDeviceController(logger);
+
+this.meetingSession = new DefaultMeetingSession(configuration, logger, deviceController, eventReporter);
+})
+```
+After in 3.0:
+
+```typescript
+const configuration = new MeetingSessionConfiguration(…);
+const eventReporter = new EventReporter(...)
+…
+const deviceController = new DefaultDeviceController(logger);
+const eventController = new DefaultEventController(configuration, logger, eventReporter)
+this.meetingSession = new DefaultMeetingSession(configuration, logger, deviceController, eventController);
+})
+```
+
+### Updating `eventDidReceive` use cases
+The `eventDidReceive` function that was part of `AudioVideoObserver` has been moved to `EventObserver` which is a observer that the `EventController` now handles. Additionally, because of this if you were you were to call `eventDidReceive` through `forEachObserver` on `AudioVideoController` you will have to call it instead on `EventController`.
+
+Before in 2.0:
+
+```typescript
+const configuration = new MeetingSessionConfiguration(…);
+…
+const deviceController = new DefaultDeviceController(logger);
+this.meetingSession = new DefaultMeetingSession(configuration, logger, deviceController);
+this.audioVideo = this.meetingSession.audioVideo;
+this.audioVideo.addObserver({
+          eventDidReceive(name: EventName, attributes: EventAttributes): void {
+            ...
+          }
+          ...
+})
+this.audioVideo.forEachObserver((observer: AudioVideoObserver) => {
+      eventDidReceive(...) 
+      ...
+    }
+);
+```
+
+After in 3.0:
+
+```typescript
+const configuration = new MeetingSessionConfiguration(…);
+…
+const deviceController = new DefaultDeviceController(logger);
+this.meetingSession = new DefaultMeetingSession(configuration, logger, deviceController);
+this.audioVideo = this.meetingSession.audioVideo;
+this.eventController = this.meetingSession.eventController;
+this.audioVideo.addObserver(...)
+this.eventController.eventController.addObserver({
+          eventDidReceive(name: EventName, attributes: EventAttributes): void {
+            ...
+          }
+});
+this.audioVideo.forEachObserver((observer: AudioVideoObserver) => {
```

Comment:
forEachObserver is more like an internal API. Wonder if we should just mark it internal. What do you think @simmkyu @devalevenkatesh

Reply:
Yes agree. The builder should get examples around the observer notification and have not to call the observers with the updates in their apps explicitly.